pkg/kv/kvserver/rangefeed/processor.go

```diff
@@ -330,7 +331,41 @@ type IntentScannerConstructor func() IntentScanner
 // for catchup-scans. Takes the key span and exclusive start time to run the
 // catchup scan for. It should be called from underneath a stopper task to
 // ensure that the engine has not been closed.
-type CatchUpIteratorConstructor func(roachpb.Span, hlc.Timestamp) (*CatchUpIterator, error)
+type CatchUpIteratorConstructor func() *CatchUpIterator
+
+// CatchUpIteratorContainer manages catchup iterator lifecycle during
+// rangefeed registration creation. Iterator is created by replica under the
+// lock to get a consistent state. Container is then passed to registration.
+// If registration is created successfully, iterator ownership is moved to
+// registration. If registration creation fails on any stage before that,
+// container could be closed to release iterator. Since ownership is moved
+// out of container, it is safe to close it regardless of startup success or
+// failure.
+type CatchUpIteratorContainer struct {
+	iter atomic.Pointer[CatchUpIterator]
+}
```

Comment:
Do we even need all of this synchronization? I think we can move the iterator construction into `registerWithRangefeedRaftMuLocked()` and pass it in directly, letting the registration assume ownership if `Processor.Register()` returns `true`, otherwise we can close the iterator on the failure path.